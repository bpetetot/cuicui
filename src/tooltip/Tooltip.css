.cc-tooltip {

}

.cc-tooltip-arrow {
  position: 'absolute';
  top: -16px;
  height: 16px;
  width: 16px;
  background-color: #333;
}

.cc-tooltip-inner {
  padding: 1em;
  background-color: #333;
  color: #fff;
  border-radius: .2em;
}