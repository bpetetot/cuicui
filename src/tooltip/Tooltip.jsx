import React, { Component } from 'react'
import PropTypes from 'prop-types'
import Popper from 'popper.js'

import TooltipOverlay from './TooltipOverlay'

class Tooltip extends Component {
  state = {
    opened: false,
    popperStyle: {},
    arrowStyle: {},
    placement: '',
  }

  componentDidMount() {
    const { placement } = this.props
    this.popperInstance = new Popper(this.targetRef, this.tooltip.popperRef, {
      placement,
      modifiers: {
        preventOverflow: { boundariesElement: 'viewport' },
        applyStyle: { enabled: false },
        updateState: {
          fn: this.updatePositions,
          enabled: true,
          order: 900,
        },
        arrow: { element: this.tooltip.arrowRef },
      },
    })
    this.popperInstance.update()
  }

  componentWillUnmount() {
    this.popperInstance.destroy()
  }

  handleHover = type => () => {
    this.setState(() => ({ opened: type === 'enter' }))
    this.popperInstance.update()
  }

  updatePositions = (data) => {
    this.setState(() => ({
      popperStyle: data.styles,
      arrowStyle: data.offsets.arrow,
      placement: data.attributes['x-placement'],
    }))
    return data
  }

  render() {
    const { tooltip, inline, children } = this.props
    return (
      <span
        ref={r => (this.targetRef = r)}
        onMouseEnter={this.handleHover('enter')}
        onMouseLeave={this.handleHover('leave')}
        style={{ display: inline ? 'inline' : 'block' }}
      >
        {children}
        <TooltipOverlay ref={t => (this.tooltip = t)} content={tooltip} {...this.state} />
      </span>
    )
  }
}

Tooltip.propTypes = {
  tooltip: PropTypes.node.isRequired,
  inline: PropTypes.bool,
  children: PropTypes.oneOfType([PropTypes.node, PropTypes.array]).isRequired,
  placement: PropTypes.oneOf(['auto', 'top', 'bottom', 'left', 'right']),
}

Tooltip.defaultProps = {
  placement: 'auto',
  inline: false,
}

export default Tooltip
