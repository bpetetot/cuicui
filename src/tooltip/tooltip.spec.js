/* eslint-disable react/jsx-filename-extension */
/* eslint-env jest */
import React from 'react'
import { mount } from 'enzyme'
import snap from 'snap'

import Tooltip from './tooltip'

jest.mock(
  'popper.js',
  () =>
    class {
      update = jest.fn(() => undefined)
      destroy = jest.fn(() => undefined)
    },
)

const snapshot = props =>
  snap(Tooltip)({ tooltip: 'my content', children: <div>Wrapped</div>, ...props })

describe('tooltip', () => {
  it('should set tooltip content', snapshot({}))
  it('should display tooltip inline', snapshot({ inline: true }))
  it('should set placement "auto"', snapshot({ placement: 'auto' }))
  it('should set placement "top"', snapshot({ placement: 'top' }))
  it('should set placement "bottom"', snapshot({ placement: 'bottom' }))
  it('should set placement "left"', snapshot({ placement: 'left' }))
  it('should set placement "right"', snapshot({ placement: 'right' }))
  it('should display on hover', () => {
    const props = { tooltip: 'my content' }
    const wrapper = mount(
      <Tooltip {...props}>
        <div>Wrapped</div>
      </Tooltip>,
    )

    wrapper.find('span').simulate('mouseEnter')
    expect(wrapper.state().opened).toBe(true)

    wrapper.find('span').simulate('mouseLeave')
    expect(wrapper.state().opened).toBe(false)
  })
  it('should update positions', () => {
    const props = { tooltip: 'my content' }
    const component = (
      <Tooltip {...props}>
        <div>Wrapped</div>
      </Tooltip>
    )
    const wrapper = mount(component)
    const inst = wrapper.instance()
    inst.updatePositions({
      styles: { position: 'absolute' },
      offsets: { arrow: { position: 'absolute' } },
      attributes: { 'x-placement': 'top' },
    })

    expect(wrapper.state()).toEqual({
      opened: false,
      popperStyle: { position: 'absolute' },
      arrowStyle: { position: 'absolute' },
      placement: 'top',
    })
  })
  it('should destroy popper on unmount', () => {
    const props = { tooltip: 'my content' }
    const component = (
      <Tooltip {...props}>
        <div>Wrapped</div>
      </Tooltip>
    )
    const wrapper = mount(component)
    const inst = wrapper.instance().popperInstance

    wrapper.unmount()

    expect(inst.destroy.mock.calls.length).toBe(1)
  })
})
