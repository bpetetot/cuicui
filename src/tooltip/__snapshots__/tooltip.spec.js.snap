// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tooltip should display tooltip inline 1`] = `
<span
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "display": "inline",
    }
  }
>
  <div>
    Wrapped
  </div>
  <div
    className="cc-tooltip"
    data-placement=""
    role="tooltip"
    style={
      Object {
        "display": "none",
      }
    }
  >
    <div
      className="cc-tooltip-arrow"
      style={Object {}}
    />
    <div
      className="cc-tooltip-inner"
    >
      my content
    </div>
  </div>
</span>
`;

exports[`tooltip should set placement "auto" 1`] = `
<span
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "display": "block",
    }
  }
>
  <div>
    Wrapped
  </div>
  <div
    className="cc-tooltip"
    data-placement=""
    role="tooltip"
    style={
      Object {
        "display": "none",
      }
    }
  >
    <div
      className="cc-tooltip-arrow"
      style={Object {}}
    />
    <div
      className="cc-tooltip-inner"
    >
      my content
    </div>
  </div>
</span>
`;

exports[`tooltip should set placement "bottom" 1`] = `
<span
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "display": "block",
    }
  }
>
  <div>
    Wrapped
  </div>
  <div
    className="cc-tooltip"
    data-placement=""
    role="tooltip"
    style={
      Object {
        "display": "none",
      }
    }
  >
    <div
      className="cc-tooltip-arrow"
      style={Object {}}
    />
    <div
      className="cc-tooltip-inner"
    >
      my content
    </div>
  </div>
</span>
`;

exports[`tooltip should set placement "left" 1`] = `
<span
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "display": "block",
    }
  }
>
  <div>
    Wrapped
  </div>
  <div
    className="cc-tooltip"
    data-placement=""
    role="tooltip"
    style={
      Object {
        "display": "none",
      }
    }
  >
    <div
      className="cc-tooltip-arrow"
      style={Object {}}
    />
    <div
      className="cc-tooltip-inner"
    >
      my content
    </div>
  </div>
</span>
`;

exports[`tooltip should set placement "right" 1`] = `
<span
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "display": "block",
    }
  }
>
  <div>
    Wrapped
  </div>
  <div
    className="cc-tooltip"
    data-placement=""
    role="tooltip"
    style={
      Object {
        "display": "none",
      }
    }
  >
    <div
      className="cc-tooltip-arrow"
      style={Object {}}
    />
    <div
      className="cc-tooltip-inner"
    >
      my content
    </div>
  </div>
</span>
`;

exports[`tooltip should set placement "top" 1`] = `
<span
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "display": "block",
    }
  }
>
  <div>
    Wrapped
  </div>
  <div
    className="cc-tooltip"
    data-placement=""
    role="tooltip"
    style={
      Object {
        "display": "none",
      }
    }
  >
    <div
      className="cc-tooltip-arrow"
      style={Object {}}
    />
    <div
      className="cc-tooltip-inner"
    >
      my content
    </div>
  </div>
</span>
`;

exports[`tooltip should set tooltip content 1`] = `
<span
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "display": "block",
    }
  }
>
  <div>
    Wrapped
  </div>
  <div
    className="cc-tooltip"
    data-placement=""
    role="tooltip"
    style={
      Object {
        "display": "none",
      }
    }
  >
    <div
      className="cc-tooltip-arrow"
      style={Object {}}
    />
    <div
      className="cc-tooltip-inner"
    >
      my content
    </div>
  </div>
</span>
`;
